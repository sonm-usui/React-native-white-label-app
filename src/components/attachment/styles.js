// @flow
import { StyleSheet, Platform } from 'react-native'
import { verticalScale, moderateScale } from 'react-native-size-matters'
import { colors, fontSizes, fontFamily } from '../../common/styles/constant'

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.white,
  },
  scrollViewWrapper: {
    backgroundColor: colors.white,
    paddingLeft: '5%',
    paddingRight: '5%',
  },
  wrapper: {
    backgroundColor: colors.white,
    paddingTop: moderateScale(12),
    ...Platform.select({
      ios: {
        borderBottomColor: colors.gray5,
        borderBottomWidth: StyleSheet.hairlineWidth,
      },
      android: {
        borderBottomColor: colors.gray5,
        borderBottomWidth: 1,
      },
    }),
  },
  title: {
    fontSize: verticalScale(fontSizes.size6),
    color: colors.gray3,
    width: '100%',
    textAlign: 'left',
    marginBottom: moderateScale(2),
    fontFamily: fontFamily,
    lineHeight: verticalScale(17),
  },
  content: {
    fontSize: verticalScale(fontSizes.size3),
    fontWeight: '700',
    color: '#505050',
    width: '100%',
    textAlign: 'left',
    fontFamily: fontFamily,
    lineHeight: verticalScale(23),
  },
  contentGray: {
    fontSize: verticalScale(fontSizes.size5),
    fontWeight: '700',
    color: colors.gray1,
    width: '100%',
    textAlign: 'left',
    fontFamily: fontFamily,
  },
  textAvatarWrapper: {
    width: '98.5%',
    flexDirection: 'row',
  },
  textWrapper: {
    width: '90%',
    paddingBottom: moderateScale(10),
  },
  avatarWrapper: {
    width: '15%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  photoAttachment: {
    width: moderateScale(150),
    height: moderateScale(150),
  },
  attachmentIcon: {
    width: moderateScale(64),
    height: moderateScale(64),
  },
  nameIconContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  attachmentNameContainer: {
    flex: 1,
    marginLeft: moderateScale(10),
  },
  parentWrapper: {
    flexDirection: 'row',
  },
  svgStyles: {
    marginRight: moderateScale(16),
  },
  extensionNameStyle: {
    fontSize: verticalScale(fontSizes.size6),
    fontWeight: '600',
    color: colors.gray1,
    width: '100%',
    textAlign: 'left',
    fontFamily: fontFamily,
  },
})
