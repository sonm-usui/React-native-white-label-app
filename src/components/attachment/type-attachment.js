// @flow
export type AttachmentPropType = {
  attachment: {
    'mime-type': string,
    data: {
      base64: string,
    },
    extension: string,
    name: string,
  },
  uid: string,
  remotePairwiseDID: string,
  label: string,
}

export type PhotoAttachmentPropType = {
  label: string,
  extension: string,
  base64: string,
}

export const attachMimeTypes = {
  photoMimeTypes: ['image/jpeg', 'image/png', 'image/jpg'],
  docMimeTypes: [
    'application/msword',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.template',
    'application/vnd.ms-word.document.macroEnabled.12',
    'application/vnd.ms-word.template.macroEnabled.12',
  ],
  excelMimeTypes: [
    'application/vnd.ms-excel',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.template',
    'application/vnd.ms-excel.sheet.macroEnabled.12',
    'application/vnd.ms-excel.template.macroEnabled.12',
    'application/vnd.ms-excel.addin.macroEnabled.12',
    'application/vnd.ms-excel.sheet.binary.macroEnabled.12',
    'text/csv',
  ],
  pptMimeTypes: [
    'application/vnd.ms-powerpoint',
    'application/vnd.openxmlformats-officedocument.presentationml.presentation',
    'application/vnd.openxmlformats-officedocument.presentationml.template',
    'application/vnd.openxmlformats-officedocument.presentationml.slideshow',
    'application/vnd.ms-powerpoint.addin.macroEnabled.12',
    'application/vnd.ms-powerpoint.presentation.macroEnabled.12',
    'application/vnd.ms-powerpoint.template.macroEnabled.12',
    'application/vnd.ms-powerpoint.slideshow.macroEnabled.12',
  ],
  pdfMimeTypes: ['application/pdf'],
  audioVideoMimeType: ['audio/mp4', 'audio/mpeg', 'audio/mp3', 'video/mp4'],
  csvMimeType: ['text/csv'],
}
