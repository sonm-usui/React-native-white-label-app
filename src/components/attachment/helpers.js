// @flow
import React from 'react'
import { Image } from 'react-native'

import { attachMimeTypes } from './type-attachment'
import { styles } from './styles'
const {
  docMimeTypes,
  csvMimeType,
  excelMimeTypes,
  pptMimeTypes,
  pdfMimeTypes,
  audioVideoMimeType,
  photoMimeTypes,
} = attachMimeTypes

export const getFileExtensionLabel = (mimeType: string) =>
  `${getFileExtensionName(mimeType)} file`

export const getFileExtensionName = (mimeType: string) => {
  switch (true) {
    case docMimeTypes.includes(mimeType):
      return 'docx'
    case csvMimeType.includes(mimeType):
      return 'CSV'
    case excelMimeTypes.includes(mimeType):
      return 'xlsx'
    case pptMimeTypes.includes(mimeType):
      return 'ppt'
    case pdfMimeTypes.includes(mimeType):
      return 'pdf'
    case audioVideoMimeType.includes(mimeType):
      return 'audio'
    case photoMimeTypes
      .filter((type) => type !== 'image/png')
      .includes(mimeType):
      return 'JPG'
    case photoMimeTypes.includes(mimeType):
      return 'PNG'
    default:
      return 'unknown'
  }
}

export const getIcon = (mimeType: string) => {
  switch (true) {
    case docMimeTypes.includes(mimeType):
      return require('../../images/docx_icon.png')
    case excelMimeTypes.includes(mimeType):
      return require('../../images/xlsx_icon.png')
    case pptMimeTypes.includes(mimeType):
      return require('../../images/ppt_icon.png')
    case pdfMimeTypes.includes(mimeType):
      return require('../../images/pdf_icon.png')
    case audioVideoMimeType.includes(mimeType):
      return require('../../images/audio_icon.png')
    default:
      return require('../../images/unknown_icon.png')
  }
}

export const checkObjectTypes = (attachment: any) =>
  !attachment['mime-type'] ||
  !attachment.data ||
  !attachment.data.base64 ||
  !attachment.extension ||
  !attachment.name
    ? true
    : false

export const PhotoAttachment = (props: { base64: string }) => {
  // TODO:KS handle error condition if image base64 encoded data is not correct
  // and image load fails
  return (
    <Image
      source={{ uri: `${props.base64}` }}
      style={styles.photoAttachment}
      resizeMode="contain"
    />
  )
}
