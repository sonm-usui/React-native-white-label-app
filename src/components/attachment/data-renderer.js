// @flow
import React, { useCallback } from 'react'
import { Text, View, Image, TouchableOpacity, Alert, Linking } from 'react-native'
import RNFetchBlob from 'rn-fetch-blob'
import FileViewer from 'react-native-file-viewer'

import type { AttachmentPropType } from './type-attachment'

import { BLANK_ATTRIBUTE_DATA_TEXT } from '../../connection-details/type-connection-details'
import { flattenAsync } from '../../common/flatten-async'
import { ExpandableText } from '../../components/expandable-text/expandable-text'

import { PhotoAttachment, checkObjectTypes, getIcon } from './helpers'
import { attachMimeTypes } from './type-attachment'
import { styles } from './styles'

const { photoMimeTypes } = attachMimeTypes

export const DataRenderer = (props: {
  label: string,
  data?: string,
  uid: string,
  remotePairwiseDID: string,
  contentStyles?: any,
}) => {
  const { label, data, uid, remotePairwiseDID, contentStyles } = props

  if (!data || data === '') {
    return (
      // Replace empty data string with (none) in lighter gray
      <Text style={[styles.contentGray, contentStyles]}>
        {BLANK_ATTRIBUTE_DATA_TEXT}
      </Text>
    )
  }

  if (label.toLowerCase().endsWith('_link')) {
    // since attribute name ends with "_link"
    // we now know that this is supposed to be an attachment
    // now, we need to render specific icon on the basis of the MIME-type
    let attachment: $PropertyType<
      AttachmentPropType,
      'attachment'
    > | null = null
    try {
      attachment = JSON.parse(data)

      // TODO:KS Harden security check around file extension
      if (checkObjectTypes(attachment)) {
        throw new Error('Invalid data')
      }
    } catch (e) {
      console.log(e.message)
      return <Text style={styles.contentGray}>Error rendering file.</Text>
    }
    // Now we know that we have got attachment data, and all fields are present
    // We can render either photo sent inside data, or we can render icon for file
    if (photoMimeTypes.includes(attachment['mime-type'].toLowerCase())) {
      return <PhotoAttachment base64={attachment.data.base64} />
    }

    return (
      <Attachment
        attachment={attachment}
        uid={uid}
        remotePairwiseDID={remotePairwiseDID}
        label={label}
      />
    )
  }
  return <ExpandableText text={data} style={[styles.content, contentStyles]} />
}

const Attachment = (props: AttachmentPropType) => {
  // This component renders icon only and not render content itself
  // MSDK is not rendering and opening items inside MSDK
  // because it is not safe from security perspective
  // Also, we don't have much time to implement, audio, video, pdf, doc, excel etc.
  // So, for file types whose mime type is known to us, we will render it's icon
  // if mime type matches word doc, then we render word doc icon

  const onKnownAttachmentOpen = useCallback(async () => {
    const {
      uid,
      remotePairwiseDID,
      label,
      attachment: {
        extension,
        data: { base64, link },
      },
    } = props
    if(props.attachment.data.link){
      Linking.openURL(props.attachment.data.link)
    } else {
    // create file path with DocumentDirectory, uid, remotePairwiseDID and label
    const attachmentPath = `${RNFetchBlob.fs.dirs.DocumentDir}/${remotePairwiseDID}-${uid}-${label}.${extension}`
    // check if file exists
    const [existError, exists] = await flattenAsync(RNFetchBlob.fs.exists)(
      attachmentPath
    )
    if (existError) {
      Alert.alert('CO001: Error opening file.')
    }
    if (!exists) {
      // if file does not exist writeFile
      const [writeError] = await flattenAsync(RNFetchBlob.fs.writeFile)(
        attachmentPath,
        base64,
        'base64'
      )

      if (writeError) {
        Alert.alert('CO002: Error opening file.')
      }
    }

    // Use file viewer to open file
    const [openError] = await flattenAsync(FileViewer.open)(attachmentPath, {
      displayName: `${label}.${extension}`,
      showOpenWithDialog: true,
      showAppsSuggestions: true,
    })
    if (openError) {
      Alert.alert(
        'CO003: Error opening file.',
        `Please install an application which can handle .${extension}`
      )
    }
    }
  }, [props])

  const icon = getIcon(props.attachment['mime-type'])

  return (
    <TouchableOpacity
      onPress={onKnownAttachmentOpen}
      style={styles.nameIconContainer}
    >
      <Image source={icon} style={styles.attachmentIcon} resizeMode="contain" />
      <View style={styles.attachmentNameContainer}>
        <Text style={styles.content} ellipsizeMode="tail" numberOfLines={3}>
          {props.attachment.name}
        </Text>
      </View>
    </TouchableOpacity>
  )
}
