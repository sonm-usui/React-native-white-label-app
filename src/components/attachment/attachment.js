// @flow
import React from 'react'
import { Text, View } from 'react-native'

import type { AttachmentPropType } from './type-attachment'

import SvgCustomIcon from '../svg-custom-icon'
import { ExpandableText } from '../expandable-text/expandable-text'
import { checkObjectTypes, getFileExtensionName } from './helpers'
import { attachMimeTypes } from './type-attachment'
import { DataRenderer } from './data-renderer'
import { styles } from './styles'

const { photoMimeTypes } = attachMimeTypes

export const RenderAttachmentIcon = (
  label: string,
  data: any,
  remotePairwiseDID?: string,
  uid?: string,
  titleStyles?: any,
  contentStyles?: any
) => {
  let attachment: $PropertyType<AttachmentPropType, 'attachment'> | null = null

  if (label.toLowerCase().endsWith('_link')) {
    try {
        attachment = JSON.parse(data)
      if (checkObjectTypes(attachment)) {
        throw new Error('Invalid data')
      }
    } catch (e) {
      console.log(e.message)
      alert('Invalid data');
      return null
    }

    return (
      <View style={styles.parentWrapper}>
        <SvgCustomIcon
          name={
            attachment !== null &&
            photoMimeTypes.includes(attachment['mime-type'].toLowerCase())
              ? 'Image'
              : 'Attachment'
          }
          style={styles.svgStyles}
          width={24}
        />
        <View style={styles.textWrapper}>
          <ExpandableText
            text={label.slice(0, -5)}
            style={[styles.title, titleStyles]}
          />
          <Text style={styles.extensionNameStyle}>
            {attachment !== null &&
              `${getFileExtensionName(attachment['mime-type'])} file`}
          </Text>
          <DataRenderer {...{ label, data, uid, remotePairwiseDID }} />
        </View>
      </View>
    )
  }

  return (
    <View style={styles.textWrapper}>
      <ExpandableText text={label} style={[styles.title, titleStyles]} />
      <DataRenderer
        {...{ label, data, uid, remotePairwiseDID, contentStyles }}
      />
    </View>
  )
}
