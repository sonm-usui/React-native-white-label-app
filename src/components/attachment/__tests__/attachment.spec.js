// @flow
import React from 'react'
import 'react-native'
import renderer from 'react-test-renderer'

import { RenderAttachmentIcon } from '../attachment'
import { DataRenderer } from '../data-renderer'

const setup = (props) => {
  const component = renderer.create(<DataRenderer {...props}></DataRenderer>)
  const instance = component.getInstance()

  return { props, component, instance }
}

describe('Render attachment icon snapshots', () => {
  it('should match snapshot', () => {
    const component = RenderAttachmentIcon(
      'label',
      'data',
      'remotePairwiseDID',
      'uid'
    )
    expect(component).toMatchSnapshot()
  })
  it('should match snapshot with icon', () => {
    const component = RenderAttachmentIcon(
      'label_link',
      '{"mime-type": "text/csv", "extension": "csv", "name": "my_csv.csv","data": { "base64": "VXNNtaXRoCgo="}}',
      'remotePairwiseDID',
      'uid'
    )
    expect(component).toMatchSnapshot()
  })
})

describe('Data renderer snapshots', () => {
  it('should match snapshot', () => {
    const { component } = setup({
      label: 'label',
      data: 'data',
      remotePairwiseDID: 'remotePairwiseDID',
      uid: 'uid',
    })
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('should match snapshot with empty data', () => {
    const { component } = setup({
      label: 'label',
      data: '',
      remotePairwiseDID: 'remotePairwiseDID',
      uid: 'uid',
    })
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(component).toMatchSnapshot()
  })
  it('should match snapshot with icon', () => {
    const { component } = setup({
      label: 'label',
      data: 'data',
      remotePairwiseDID:
        '{"mime-type": "text/csv", "extension": "csv", "name": "my_csv.csv","data": { "base64": "VXNNtaXRoCgo="}}',
      uid: 'uid',
    })
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
