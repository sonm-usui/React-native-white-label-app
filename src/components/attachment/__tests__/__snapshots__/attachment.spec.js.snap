// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data renderer snapshots should match snapshot 1`] = `
<Text
  ellipsizeMode="tail"
  numberOfLines={2}
  onTextLayout={[Function]}
  style={
    Array [
      Object {
        "color": "#505050",
        "fontFamily": "System",
        "fontSize": 15.088235294117647,
        "fontWeight": "700",
        "lineHeight": 18.26470588235294,
        "textAlign": "left",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  data
</Text>
`;

exports[`Data renderer snapshots should match snapshot with empty data 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#505050",
        "fontFamily": "System",
        "fontSize": 11.911764705882351,
        "fontWeight": "700",
        "textAlign": "left",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  n/a
</Text>
`;

exports[`Data renderer snapshots should match snapshot with empty data 2`] = `
<Text
  style={
    Array [
      Object {
        "color": "#505050",
        "fontFamily": "System",
        "fontSize": 11.911764705882351,
        "fontWeight": "700",
        "textAlign": "left",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  n/a
</Text>
`;

exports[`Data renderer snapshots should match snapshot with icon 1`] = `
<Text
  ellipsizeMode="tail"
  numberOfLines={2}
  onTextLayout={[Function]}
  style={
    Array [
      Object {
        "color": "#505050",
        "fontFamily": "System",
        "fontSize": 15.088235294117647,
        "fontWeight": "700",
        "lineHeight": 18.26470588235294,
        "textAlign": "left",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  data
</Text>
`;

exports[`Render attachment icon snapshots should match snapshot 1`] = `
<View
  style={
    Object {
      "paddingBottom": 9.571428571428571,
      "width": "90%",
    }
  }
>
  <ExpandableText
    style={
      Array [
        Object {
          "color": "#A5A5A5",
          "fontFamily": "System",
          "fontSize": 11.117647058823529,
          "lineHeight": 13.5,
          "marginBottom": 1.9142857142857141,
          "textAlign": "left",
          "width": "100%",
        },
        undefined,
      ]
    }
    text="label"
  />
  <DataRenderer
    data="data"
    label="label"
    remotePairwiseDID="remotePairwiseDID"
    uid="uid"
  />
</View>
`;

exports[`Render attachment icon snapshots should match snapshot with icon 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <SvgCustomIcon
    name="Attachment"
    style={
      Object {
        "marginRight": 15.314285714285713,
      }
    }
    width={24}
  />
  <View
    style={
      Object {
        "paddingBottom": 9.571428571428571,
        "width": "90%",
      }
    }
  >
    <ExpandableText
      style={
        Array [
          Object {
            "color": "#A5A5A5",
            "fontFamily": "System",
            "fontSize": 11.117647058823529,
            "lineHeight": 13.5,
            "marginBottom": 1.9142857142857141,
            "textAlign": "left",
            "width": "100%",
          },
          undefined,
        ]
      }
      text="label"
    />
    <Text
      style={
        Object {
          "color": "#505050",
          "fontFamily": "System",
          "fontSize": 11.117647058823529,
          "fontWeight": "600",
          "textAlign": "left",
          "width": "100%",
        }
      }
    >
      CSV file
    </Text>
    <DataRenderer
      data="{\\"mime-type\\": \\"text/csv\\", \\"extension\\": \\"csv\\", \\"name\\": \\"my_csv.csv\\",\\"data\\": { \\"base64\\": \\"VXNNtaXRoCgo=\\"}}"
      label="label_link"
      remotePairwiseDID="remotePairwiseDID"
      uid="uid"
    />
  </View>
</View>
`;
